src/pkcs11/openssl.c

```diff
@@ -419,7 +421,38 @@ CK_RV sc_pkcs11_verify_data(const unsigned char *pubkey, int pubkey_len,
 #endif
 	}
 
-	pkey = d2i_PublicKey(EVP_PKEY_RSA, NULL, &pubkey, pubkey_len);
+	/*
+	 * PKCS#11 does not define CKA_VALUE for public keys, and different cards
+	 * return either the raw or spki versions as defined in PKCS#15
+	 * And we need to support more then just RSA.
+	 * We can use d2i_PUBKEY which works for SPKI and any key type. 
+	 */
+	pubkey_tmp = pubkey; /* pass in so pubkey pointer is not modified */
+	/* SPKI ASN.1 starts with SEQUENCE, SEQUENCE, OBJECT IDENTIFIER */
+	if (*pubkey_tmp++ == 0x30) {
+		if (*pubkey_tmp & 0x80)
+			pubkey_tmp += *pubkey_tmp & 0x7F;
+
+		pubkey_tmp++;
+		if (*pubkey_tmp++ == 0x30) {
+			if (*pubkey_tmp & 0x80)
+				pubkey_tmp += *pubkey_tmp & 0x7F;
+
+			pubkey_tmp += 1;
+			if (*pubkey_tmp == 0x06)
+				is_spki = 1;
+		}
+	}
```

Comment:
I like the Idea of determining the type beforehand. However, you should at least use the ASN.1 parsing routines of OpenSC. Even better, you should check whether you can reuse the [existing code that parses this kind of data](https://github.com/OpenSC/OpenSC/blob/e7915ec198a4f2861141157b8fcd6e4e1322eaff/src/libopensc/pkcs15-cert.c#L62)